SofaKernel/modules/SofaBaseMechanics/src/SofaBaseMechanics/DiagonalMass.inl

```diff
@@ -1219,15 +1353,24 @@ void DiagonalMass<DataTypes, MassType>::initFromVertexMass()
 {
     msg_info() << "vertexMass information is used";
 
-    const MassVector& vertexMass = d_vertexMass.getValue();
+    // save a copy of input vertexMass vector
+    MassVector vertexMass = d_vertexMass.getValue();
     Real totalMassSave = 0.0;
     for(size_t i=0; i<vertexMass.size(); i++)
     {
         totalMassSave += vertexMass[i];
     }
 
+    // set total mass
     d_totalMass.setValue(totalMassSave);
-    initFromTotalMass();
+
+    // compute vertexMass vector with density == 1
+    Real sumMass = computeVertexMass(1.0);
+
+    // Set real density from sumMass found
+    setMassDensity(Real(totalMassSave / sumMass));
```

Comment:
If computeVertexMass() returns 0 (for example because of no topology), this division wont be pretty here 🤐

Reply:
the method setMassDensity has a check value : https://github.com/sofa-framework/sofa/blob/86009f06b3d47c0d185f58d0a47c6021124f5895/SofaKernel/modules/SofaBaseMechanics/src/SofaBaseMechanics/DiagonalMass.inl#L1383-L1397